File: code/data_structures/src/queue/queue_using_stack/queue_using_stack.cpp
#include <iostream>
#include <stack>
using namespace std;
// Part of Cosmos by OpenGenus Foundation
// queue data structure using two stacks
class queue {
private:
    stack<int> s1, s2;
public:
    void enqueue(int element);
    int dequeue();
    void displayQueue();
};

// enqueue an element to the queue
void queue :: enqueue(int element)
{
    s1.push(element);
}

// dequeue the front element
int queue :: dequeue()
{
    // transfer all elements of s1 into s2
    if (s1.empty() && s2.empty())
        return 0;

    while (!s1.empty())
    {
        s2.push(s1.top());
        s1.pop();
    }
    // pop and store the top element from s2

    int ret = s2.top();
    s2.pop();
    // transfer all elements of s2 back to s1
    while (!s2.empty())
    {
        s1.push(s2.top());
        s2.pop();
    }
    return ret;
}

//display the elements of the queue
void queue :: displayQueue()
{
    cout << "\nDisplaying the queue :-\n";
    while (!s1.empty())
    {
        s2.push(s1.top());
        s1.pop();
    }
    while (!s2.empty())
    {
        cout << s2.top() << " ";
        s1.push(s2.top());
        s2.pop();
    }
}

// main
int main()
{
    queue q;
    int exit = 0;
    int enqueue;
    char input;
    while (!exit)
    {
        cout << "Enter e to enqueue,d to dequeue,s to display queue,x to exit: ";
        cin >> input;
        if (input == 'e')
        {
            cout << "Enter number to enqueue: ";
            cin >> enqueue;
            q.enqueue(enqueue);
        }

        if (input == 'd')
            q.dequeue();

        if (input == 's')
        {
            q.displayQueue();
            cout << endl;
        }
        if (input == 'x')
            break;
    }

    return 0;
}

File: code/data_structures/src/queue/queue_using_stack/queue_using_stack.py
# Part of Cosmos by OpenGenus Foundation


class Stack:
    def __init__(self):
        self.items = []

    def push(self, x):
        self.items.append(x)

    def pop(self):
        return self.items.pop()

    def size(self):
        return len(self.items)

    def is_empty(self):
        return self.size() == 0


class Queue:
    def __init__(self):
        self.s1 = Stack()
        self.s2 = Stack()

    def enqueue(self, p):
        self.s1.push(p)

    def dequeue(self):
        # pop all elements in s1 onto s2
        while not self.s1.is_empty():
            self.s2.push(self.s1.pop())

        # pop all elements of s1
        return_value = self.s2.pop()

        # re pop all elements of s2 onto s1
        while not self.s2.is_empty():
            self.s1.push(self.s2.pop())

        return return_value


if __name__ == "__main__":
    from random import choice

    a = Queue()

    l = [choice(range(1, 500)) for i in range(1, 10)]
    print(l)

    for i in l:
        a.enqueue(i)

    for i in range(len(l)):
        print("Dequeued:", a.dequeue())
